plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,391 @@
+    """
+    is_horizontal = (direction == 'horizontal')
+    new_top_left_x = top_left_x + is_horizontal*0.5*box_w
+    new_top_left_y = top_left_y + (not is_horizontal)*0.5*box_h
+    new_box_w = (not is_horizontal)*box_w + is_horizontal
+    new_box_h = (not is_horizontal) + is_horizontal*box_h
+
+    html_box = """<!-- Draw some lines in -->
+          context.beginPath();
+          context.rect({top_left_x}, {top_left_y}, {box_w}, {box_h});
+          context.lineWidth = 1;
+          context.strokeStyle = 'black';
+          context.stroke();
+    """.format(top_left_x=new_top_left_x, top_left_y=new_top_left_y,
+               box_w=new_box_w, box_h=new_box_h)
+
+    index_for_new_box = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_for_new_box] + html_box +
+                      dashboard_html[index_for_new_box:])
+    return dashboard_html
+
+
+def add_html_text(dashboard_html, text, top_left_x, top_left_y, box_w, box_h):
+    """
+    Add a number to the middle of an HTML box.
+    """
+    html_text = """<!-- Insert box numbers -->
+          context.font = '{font_size}pt Times New Roman';
+          context.textAlign = 'center';
+          context.fillText({text}, {top_left_x} + 0.5*{box_w}, {top_left_y} + 0.5*{box_h});
+    """.format(text=text, top_left_x=top_left_x, top_left_y=top_left_y,
+               box_w=box_w, box_h=box_h, font_size=font_size)
+
+    index_to_add_text = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_to_add_text] + html_text +
+                      dashboard_html[index_to_add_text:])
+    return dashboard_html
+
+
+class Dashboard(dict):
+    def __init__(self, content=None):
+        if content is None:
+            content = {}
+
+        self.box_ids_to_path = {}
+        if not content:
+            self['layout'] = _empty_box()
+            self['version'] = 2
+            self['settings'] = {}
+        else:
+            self['layout'] = content['layout']
+            self['version'] = content['version']
+            self['settings'] = content['settings']
+
+            self._assign_boxes_to_ids()
+
+    def _assign_boxes_to_ids(self):
+        self.box_ids_to_path = {}
+        all_nodes = []
+        node_gen = node_generator(self['layout'])
+
+        finished_iteration = False
+        while not finished_iteration:
+            try:
+                all_nodes.append(node_gen.next())
+            except StopIteration:
+                finished_iteration = True
+
+        for node in all_nodes:
+            if (node[1] != () and node[0]['type'] == 'box'
+                    and node[0]['boxType'] != 'empty'):
+                try:
+                    max_id = max(self.box_ids_to_path.keys())
+                except ValueError:
+                    max_id = 0
+                self.box_ids_to_path[max_id + 1] = list(node[1])
+
+    def _insert(self, box_or_container, path):
```

Comment:
1. and 2. is _really_ confusing and non-intuitive for me. It's funny how two people can be on both sides of this where the intuitive thing for one is not intuitive for the other, but anyways...

I don't like how in 1. we're saying "A `box` is a `CONTAINER`" because the object with `type: 'split'` that holds a `first` and `second` has the duck-like (am I using that right?) to something "that contains". Calling a box a container just violates my idea of what a container should be and what properties, namely having children in `first` and `second`, a container should have.

I get the logic in 3. and 4.; `plotBox` and `emptyBox` are subsets of a `box`, specifying the type.

What do you think about this? Alternatively, I could switch up the terminology from `container` to `split` (so `box_or_container` becomes `box_or_split`)